### DISCUSSION TITLE
[Android] Crowdin CLI tool could not find pre-fetched source files

### DISCUSSION THREAD

**Discussion by Afolayan** [crowdin/crowdin-cli#464] [Q&A] [answered]
**Describe the bug**
I am currently working on setting up Crowdin cli for a multi-module project for Android.I've followed the directions on how to pull sources and translations with the use of the cli-tool, but recently encountered an issue. Currently using the Crowdin-figma plugin, hence do not have source files in the repository, and the need to pull sources and their respective translations. 

**To Reproduce**
I have setup the config and a kotlin script to move source files to their right destinations.

1. Configuration file: the config for pulling translations
The repo has been setup to follow this format: `feature/$featureName/src/main...`. The source file exists in the `values` folder. The string file for each of the feature is named as `$featureName.xml`. This is so that config can be reusable for multiple android modules.
```
{
    "source": "/feature/*/src/main/res/values/*.*",
    "translation": "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "skip_untranslated_files": true,
    "export_only_approved": true,
 }
```
I have this configured in my Crowdin project, 
`Resulting file after translations export:`
`/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%`

2. Execute CLI command: `crowdin download --identity {{crowdinCredentials}} --config {{translationsConfig}} --verbose`

3. See an error
<img width="1579" alt="Screenshot 2022-08-04 at 01 37 02" src="https://user-images.githubusercontent.com/9561500/182730796-3778491b-a1cf-4bfe-8fb2-e7d31e84b786.png">

**Expected behavior**
It does well to output the correct folder structure for the languages available, but it could not place the translations file in those folders.
Am I doing something wrong? 

**Environment:**
 - OS: MacOS
 - Version: 3.7.9

**Screenshots**

**Additional context**

**Comment by andrii-bodnar** [comment]
Hi @Afolayan!

Just checked your configuration on a test project and can't reproduce this issue.

Config:
![config](https://user-images.githubusercontent.com/29282228/182783635-ac15eb7a-faf4-4b1e-8947-665d81c666a6.png)

File export pattern in Crowdin:
![export pattern](https://user-images.githubusercontent.com/29282228/182784160-6f8689a4-375f-4eff-b735-435ff0eecb7a.png)

CLI output:
![cli output](https://user-images.githubusercontent.com/29282228/182783746-1d7fbaa1-0eed-4e54-bba9-7ffa99cce446.png)

Files structure after translations downloading:
![files structure](https://user-images.githubusercontent.com/29282228/182783797-abe138f9-4a5e-497d-8a80-a8348ff898f1.png)

Could you please provide more details on how to reproduce this issue? Do you have a Language Mapping configured in your Crowdin project?

**Comment by Afolayan** [comment]
Hi @andrii-bodnar thanks for looking into this.
This is the full config block:
```
{
    "source": "feature/*/values/*.*",
    "translation": "feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "skip_untranslated_files": true,
    "export_only_approved": true,
    "translation_replace": {
        "pt_BR": "pt",
        "zh-CN": "zh-rCN",
        "zh-TW": "zh-rTW",
        "en_": ""
    },
    "excluded_target_languages": [
      "ach"
    ]
 }
```

I tried removing `translation_replace` and `excluded_target_languages` sections from the config, I got the same error.
I also changed my command from `crowdin download` to `crowdin pull`, yet the same error.

The project is new and the translations folder (`values-%osx_locale%`) do not yet exist. Tried creating empty folders but didn't work out.

Did you run any command before running `crowdin pull`?

**Comment by andrii-bodnar** [comment]
The `crowdin pull` command is just an alias for the `crowdin download`

Before this command, I ran only the `crowdin push` command to upload mt source file to Crowdin

I will try adding the `translation_replace` option to my config

**Comment by andrii-bodnar** [comment]
@Afolayan please also check the language mapping configuration in your Crowdin project settings. There might be some mappings that could affect the translations download

![screenshot-crowdin com-2022 08 04-10_32_54](https://user-images.githubusercontent.com/29282228/182790473-ea763c21-7cbe-4040-ac16-38ec064d89a6.png)

**Comment by Afolayan** [comment]
I have a language mapping setup. This is the same as the one in config.

<img width="492" alt="Screenshot 2022-08-04 at 09 36 13" src="https://user-images.githubusercontent.com/9561500/182791023-b1a6651a-ee9a-4b67-8519-cb79eb5944b9.png">

**Comment by andrii-bodnar** [comment]
@Afolayan thanks for the details provided!

Also, I noticed the following options in your configuration file:

```
    "skip_untranslated_files": true,
    "export_only_approved": true,
```

Please make sure that your files are fully translated and approved.

**Comment by Afolayan** [comment]
I removed all the extra config lines and was left with this:
```
files: [
 {
    "source": "/feature/*/src/main/res/values/*.*",
    "translation": "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
 },
]
```
No luck.

Just to be sure:
`source` is the location of the source file on the local machine, while
`translation` is the same value in the *Resulting file after translations export* section of the file settings on Crowdin.
Is this correct?

**Comment by andrii-bodnar** [comment]
Yes, it's correct.

Could you please provide an example of the expected files and folder structure after the translations downloading?

I mean which language codes are you expecting?

> No luck.

Please provide the CLI output

**Comment by Afolayan** [comment]
I currently have the `track.xml` file in the listed languages.

<img width="1622" alt="Screenshot 2022-08-04 at 10 00 25" src="https://user-images.githubusercontent.com/9561500/182795564-eb283eff-1e00-4906-80f0-f6ddb85d4fd1.png">

**Comment by andrii-bodnar** [comment]
@Afolayan could you please share your Crowdin project ID?

**Comment by andrii-bodnar** [comment]
Do you have the `base_path` option in your CLI configuration?

**Comment by Afolayan** [comment]
Yes I have `base_path`
```

"base_path" : "."
"base_url" : "https://api.crowdin.com"

files: [
 {
    "source": "/feature/*/src/main/res/values/*.*",
    "translation": "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
 },
]
```

Sharing the project ID? Isn't that a security issue? Can I send it to an email or somewhere much private?

**Comment by andrii-bodnar** [comment]
@Afolayan please use [this form](https://crowdin.com/contacts) and mention the current issue link.

**Comment by Afolayan** [comment]
Sent.

**Comment by andrii-bodnar** [comment]
@Afolayan I still can't reproduce this issue.

![config](https://user-images.githubusercontent.com/29282228/182808143-04a789e8-cf7e-4b8b-88a8-d2b894e0228d.png)

![settings](https://user-images.githubusercontent.com/29282228/182808154-4e7fd616-a851-4ca2-b322-911069ec5836.png)

![output](https://user-images.githubusercontent.com/29282228/182808172-e7f61f97-3e83-4678-a220-323854f99963.png)

Do you use a single configuration file? Please provide the exact steps to reproduce and the full version of CLI configuration (without credentials)

**Comment by Afolayan** [comment]
I have two config files: one to pull the sources, the other to download the translations
Source config: (this worked, i could see the downloaded source file 
```
"base_path" : "."
"base_url" : "https://api.crowdin.com"

"preserve_hierarchy": true

files: [
{
    "source" : "%original_file_name%",
    "translation" : "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "export_only_approved": true,
    "skip_untranslated_files": true,
    "translation_replace": {
        "pt_BR": "pt",
        "zh-CN": "zh-rCN",
        "zh-TW": "zh-rTW",
        "en_": ""
    },
    "ignore": [
    "clue-android.tbx",
    "features.json"
    ],
    "excluded_target_languages": [
        "ach"
    ]
 }
]
```

Config to pull/download translations
```
"base_path" : "."
"base_url" : "https://api.crowdin.com"

files: [
 {
    "source": "/feature/*/src/main/res/values/*.*",
    "translation": "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "skip_untranslated_files": true,
    "export_only_approved": true,
    "translation_replace": {
        "pt_BR": "pt",
        "zh-CN": "zh-rCN",
        "zh-TW": "zh-rTW",
        "en_": ""
    },
    "excluded_target_languages": [
      "ach"
    ]
 },
]
```

The folder structure for the android project goes like
```
- feature/
      - track
            - src/main/res/values/
      - {{other modules}}
```

I hope this is sufficient.

**Comment by andrii-bodnar** [comment]
@Afolayan thanks for the details provided. 

Which config is used for the **sources upload** to Crowdin?

Also, please provide a clear sequence of CLI runs using these configuration files.

**Comment by Afolayan** [comment]
There's no config for uploading sources. We use the Crowdin-Figma plugin to populate the file on Crowdin. So the sequence of steps in the script only pulls the sources and the translations.

```
1. crowdin download sources --identity $crowdinCredentials --config $sourceConfig --verbose
2. crowdin download --identity $crowdinCredentials --config $translationsConfig --verbose
```

**Comment by Afolayan** [comment]
The config at Config to pull/download translations:
```
"base_path" : "."
"base_url" : "https://api.crowdin.com"

files: [
 {
    "source": "/feature/*/src/main/res/values/*.*",
    "translation": "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "skip_untranslated_files": true,
    "export_only_approved": true,
    "translation_replace": {
        "pt_BR": "pt",
        "zh-CN": "zh-rCN",
        "zh-TW": "zh-rTW",
        "en_": ""
    },
    "excluded_target_languages": [
      "ach"
    ]
 },
]
```

**Comment by andrii-bodnar** [comment]
@Afolayan it looks like you're using the first configuration file instead of using the second file for translations download.

**Comment by andrii-bodnar** [comment]
I was able to reproduce this error by passing this file as a `--config` parameter for `crowdin download`:

```
"base_path" : "."
"base_url" : "https://api.crowdin.com"

"preserve_hierarchy": true

files: [
{
    "source" : "%original_file_name%",
    "translation" : "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "export_only_approved": true,
    "skip_untranslated_files": true,
    "translation_replace": {
        "pt_BR": "pt",
        "zh-CN": "zh-rCN",
        "zh-TW": "zh-rTW",
        "en_": ""
    },
    "ignore": [
    "clue-android.tbx",
    "features.json"
    ],
    "excluded_target_languages": [
        "ach"
    ]
 }
]
```

and this is expected for this config.

**Comment by Afolayan** [comment]
Hi @andrii-bodnar, is there a single config I can use to download the sources and translations, instead of having two config files?

Also, I converted my script from bash to kotlin sometimes in the past. I went back to the bash script and I was able to pull the source and translation files. Could this be responsible? The kotlin script does the same thing, just the change of language.

**Comment by andrii-bodnar** [comment]
@Afolayan if it's working with your bash script and not working with the Kotlin script then there is some bug in the Kotlin script - probably you're passing the wrong CLI configuration file to the second command (translations download).

**Comment by andrii-bodnar** [comment, marked as answer]
@Afolayan here is a working single config that will allow you to download source and translation files:

```yml
"preserve_hierarchy": "true"

files: [
 {
    "source": "/feature/*/src/main/res/values/*.*",
    "translation": "/feature/%file_name%/src/main/res/values-%osx_locale%/%original_file_name%",
    "skip_untranslated_files": true,
    "export_only_approved": true,
    "excluded_target_languages": [
      "ach"
    ]
 },
]
```

You don't need the `translation_replace` here. Just configure the right language mapping in your Crowdin project settings.

It requires **changes in the files structure** in the Crowdin project (just create the folders in UI and move the file):

![files](https://user-images.githubusercontent.com/29282228/182876237-4ff97463-b1b6-4e7b-846f-9c888c401207.png)

And my CLI output:

![output](https://user-images.githubusercontent.com/29282228/182876332-8fa9e289-aa53-4d82-9c46-3606103d744e.png)

As you can see, I was able either to download the source and translations.

**Comment by Afolayan** [comment]
Thanks a lot @andrii-bodnar for assisting with this. 
I wish there's another solution that does not require updating the file structure on Crowdin. I feel it may be a lot of work for our translators to navigate the empty folders. I'll be using the two config files while I look out for better solution.

I removed the `translation_replace` field and cleaned up the config files. 
There was a little issue with the Kotlin script; it's been updated now and all looks good from my end.

**Comment by andrii-bodnar** [comment]
Hi @Afolayan!
Happy to hear that it's working for you now! Thanks for the update
